### PR TITLE
Fix mem pool restore mem block cause slow creation

### PR DESCRIPTION
**Pull request by zeroqn**

### CODE REVIEW

File: crates/rpc-server/src/registry.rs [hunk 1]
```diff
@@ -339,6 +339,30 @@ impl RequestSubmitter {
     }
 
     async fn in_background(self) {
+        // First mem pool reinject txs
+        {
+            let db = self.store.begin_transaction();
+            let mut mem_pool = self.mem_pool.lock().await;
+
+            while let Some(hash) = mem_pool.reinject_txs_mut().last().cloned() {
```
Comment (jjyr) [inline crates/rpc-server/src/registry.rs]:
```suggestion
            while let Some(hash) = mem_pool.reinject_txs_mut().pop() {
```

File: crates/mem-pool/src/pool.rs [hunk 2]
```diff
@@ -230,6 +205,10 @@ impl MemPool {
         self.mem_block.txs().len().saturating_add(expect_slots) > MAX_MEM_BLOCK_TXS
     }
 
+    pub fn reinject_txs_mut(&mut self) -> &mut Vec<H256> {
```
Comment (jjyr) [inline crates/mem-pool/src/pool.rs]:
```suggestion
    pub fn pending_restored_tx_hashes(&mut self) -> &mut Vec<H256> {
```

The name should be meaningful

File: crates/mem-pool/src/mem_block.rs [hunk 3]
```diff
@@ -131,6 +131,20 @@ impl MemBlock {
         self.state_checkpoints.push(state_checkpoint);
     }
 
+    pub fn emergency_reinject_restored_tx_hashes(&mut self, tx_hashes: &[H256]) {
```
Comment (jjyr) [inline crates/mem-pool/src/mem_block.rs]:
```suggestion
    pub fn force_reinject_tx_hashes(&mut self, tx_hashes: &[H256]) {
```

File: crates/mem-pool/src/mem_block.rs [hunk 4]
```diff
@@ -131,6 +131,20 @@ impl MemBlock {
...
+            if !self.txs_set.contains(tx_hash) {
+                self.txs_set.insert(*tx_hash);
+                self.txs.push(*tx_hash);
+            }
+        }
+    }
+
+    pub fn clear_txs(&mut self) {
+        self.txs_set.clear();
+        self.txs.clear();
```
Comment (jjyr) [inline crates/mem-pool/src/mem_block.rs:157]: We should also clear `state_checkpoints` & `touched_keys`

**Comment by jjyr** [comment]
pls add some tests and send pr to develop branch

**Comment by zeroqn** [comment]
> pls add some tests and send pr to develop branch

Done

File: crates/mem-pool/src/restore_manager.rs [hunk 5]
```diff
@@ -112,7 +115,7 @@ impl RestoreManager {
             None => return Ok(None),
         };
 
-        let block = packed::MemBlock::from_slice(&read(file_path)?)?;
+        let block = packed::CompactMemBlock::from_slice(&read(file_path)?)?;
```
Comment (jjyr) [inline crates/mem-pool/src/restore_manager.rs]: Does this compatible this the current version?

File: crates/rpc-server/src/registry.rs [hunk 6]
```diff
@@ -339,6 +339,32 @@ impl RequestSubmitter {
...
+        {
+            let db = self.store.begin_transaction();
+            let mut mem_pool = self.mem_pool.lock().await;
+
+            log::info!(
+                "reinject mem block txs {}",
+                mem_pool.pending_restored_tx_hashes().len()
+            );
+            while let Some(hash) = mem_pool.pending_restored_tx_hashes().pop() {
+                match db.get_mem_pool_transaction(&hash) {
```
Comment (jjyr) [inline crates/rpc-server/src/registry.rs:367]: The order of txs is reverse? This may cause nonce error.

**Review by jjyr — CHANGES_REQUESTED** [review]

File: crates/mem-pool/src/restore_manager.rs (hunk 5 above)
Comment (zeroqn) [reply crates/mem-pool/src/restore_manager.rs]: Fixed

File: crates/rpc-server/src/registry.rs (hunk 6 above)
Comment (zeroqn) [reply crates/rpc-server/src/registry.rs:367]: Indeed. Fixed.

**Review by jjyr — DISMISSED** [review]

File: crates/types/src/offchain/extension.rs [hunk 7]
```diff
@@ -56,3 +56,22 @@ pub fn global_state_from_slice(slice: &[u8]) -> Result<GlobalState, Verification
...
+        CompactMemBlock::new_builder()
+            .txs(block.txs())
+            .withdrawals(block.withdrawals())
+            .deposits(block.deposits())
+            .build()
+    }
+}
+
+impl CompactMemBlock {
+    pub fn from_full_compitablie_slice(slice: &[u8]) -> Result<CompactMemBlock, VerificationError> {
```
Comment (Flouse) [inline crates/types/src/offchain/extension.rs]: rename it to from_full_compatible_slice ?

**Review by Flouse — DISMISSED** [review]

File: crates/types/src/offchain/extension.rs (hunk 7 above)
Comment (zeroqn) [reply crates/types/src/offchain/extension.rs]: Fix typo

**Review by jjyr — APPROVED** [review]

**Review by Flouse — APPROVED** [review]

**Pull request merged**